lib/wasi-types/src/io.rs

```diff
@@ -0,0 +1,19 @@
+use wasmer_types::ValueType;
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[repr(C)]
+pub struct __wasi_ciovec_t {
+    pub buf: u32,
```

Comment:
This changed from `WasmPtr<u8, Array>` to `u32` to remove the need to use `WasmPtr` in `wasmer-wasi-types`. But I believe it's complicated to create this type now. Imagine you have the following code:

```rust
let mut vec: Vec<u8> = vec![0; 2];
let iov = __wasi_ciovec_t {
    buf: vec.as_mut_ptr(),
    buf_len: vec.len() as u32,
};
```

it fails to compile because `buf` is a `u32`, not a `*mut u8`. Casting a pointer to a usize is unstable for the moment (it requires the `const_raw_ptr_to_usize_cast` feature). It's not ideal.

Any recommandation @MarkMcCaskey?

Reply:
A workaround is to:

```rust
    buf: vec.as_mut_ptr() as usize as u32,
```

I'm not super happy with that though.

Reply:
> Casting a pointer to a usize is unstable for the moment 

Really? I've done it all the time and never seen this. I wonder if it's because it's `*mut`?  🤔 

I think your work around is what I'd expect here. We can have a `WasmUsize` type or something, but even that is confusing because with Wasm64, Wasm will have both 64 and 32 bit pointers depending on context. I'm unsure to what extent that will happen in practice given the target triples are all `wasm32-*`, maybe it'll be fully separated into `wasm64-*` and `wasm32-*`.

Reply:
It's out of the scope of this PR, but yeah, I agree that we must use a type alias to represent the memory size, something like `wasm_usize`/`WasmUsize` is an excellent idea.